### ISSUE TITLE
[VarExporter] Error in LazyGhostTrait with Symfony 6.4.0

### ISSUE THREAD

**Issue by dmaicher** [symfony/symfony#52793]
### Symfony version(s) affected

6.4.0

### Description

Ran into this issue today while running a Symfony command that uses a service marked as `lazy: true`

![Screenshot from 2023-11-29 12-49-11](https://github.com/symfony/symfony/assets/921145/a97d72a6-f518-44d0-8113-75ddc171bc15)

### How to reproduce

I don't have a reproducer yet :confused: ... will try to come up with something later today.

### Additional Context

Seems to be caused by https://github.com/symfony/symfony/pull/52762/files potentially? 

cc @nicolas-grekas

**Comment by dmaicher** [comment]
So this happens for this class/service which is marked as `lazy: true`:

```php
class QueuePublisher implements QueuePublisherInterface
{
    /**
     * @var array<string, TransportInterface>
     */
    private array $transports = [];

    public function publish(string $queueName, Message $message): void
    {
        isset($this->transports[$queueName]);
    }
}
```

If I call `$instance->publish(...)` then it fails on line `isset($this->transports[$queueName])` with

![Screenshot from 2023-11-29 12-49-11](https://github.com/symfony/symfony/assets/921145/0e322cae-d04d-4f5f-954a-4677aeb7ab12)

Somehow the `$transports` array member is causing trouble?

**Comment by talha-core** [comment]
I got the same issue.... 

In my Entity, I have this:
```php
/** @var Collection<int, ArticlePrice> */
#[ORM\OneToMany(targetEntity: ArticlePrice::class, mappedBy: 'articleCombine', cascade: ['persist'])]
#[ORM\OrderBy(['type' => 'ASC', 'qty' => 'ASC', 'id' => 'ASC'])]
private Collection $articlePriceOptions;
```

And I have this function in the same entity:
```php
public function getArticlePriceConsumerOptions(): Collection
{
    return $this->articlePriceOptions->filter(
        function (ArticlePrice $articlePrice) {
            return $articlePrice->getType() === 'c';
        }
    );
}
```

But I am getting the same error:

```
ErrorException {#7361 ▼
  #message: "Warning: Undefined array key 3"
  #code: 0
  #file: "[/var/www/vendor/symfony/var-exporter/LazyGhostTrait.php](file:///var/www/vendor/symfony/var-exporter/LazyGhostTrait.php#L170)"
  #line: 170
  #severity: E_WARNING
  trace: {▼
    [/var/www/vendor/symfony/var-exporter/LazyGhostTrait.php:170](file:///var/www/vendor/symfony/var-exporter/LazyGhostTrait.php#L170) {▶}
    [/var/www/src/Entity/ArticleCombine.php:277](file:///var/www/src/Entity/ArticleCombine.php#L277) {▶}
    [/var/www/src/Twig/ShopTwigExtension.php:300](file:///var/www/src/Twig/ShopTwigExtension.php#L300) {▶}
    [/var/www/vendor/twig/twig/src/Environment.php:360](file:///var/www/vendor/twig/twig/src/Environment.php#L360) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:394](file:///var/www/vendor/twig/twig/src/Template.php#L394) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:367](file:///var/www/vendor/twig/twig/src/Template.php#L367) {▶}
    [/var/www/vendor/twig/twig/src/Environment.php:360](file:///var/www/vendor/twig/twig/src/Environment.php#L360) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:394](file:///var/www/vendor/twig/twig/src/Template.php#L394) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:367](file:///var/www/vendor/twig/twig/src/Template.php#L367) {▶}
    [/var/www/vendor/twig/twig/src/Environment.php:360](file:///var/www/vendor/twig/twig/src/Environment.php#L360) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:394](file:///var/www/vendor/twig/twig/src/Template.php#L394) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:367](file:///var/www/vendor/twig/twig/src/Template.php#L367) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:379](file:///var/www/vendor/twig/twig/src/Template.php#L379) {▶}
    [/var/www/vendor/twig/twig/src/TemplateWrapper.php:38](file:///var/www/vendor/twig/twig/src/TemplateWrapper.php#L38) {▶}
    [/var/www/vendor/twig/twig/src/Environment.php:280](file:///var/www/vendor/twig/twig/src/Environment.php#L280) {▶}
    [/var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php:448](file:///var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php#L448) {▶}
    [/var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php:453](file:///var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php#L453) {▶}
    [/var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php:253](file:///var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php#L253) {▶}
    [/var/www/src/Controller/HeadController.php:57](file:///var/www/src/Controller/HeadController.php#L57) {▶}
    [/var/www/vendor/symfony/http-kernel/HttpKernel.php:181](file:///var/www/vendor/symfony/http-kernel/HttpKernel.php#L181) {▶}
    [/var/www/vendor/symfony/http-kernel/HttpKernel.php:76](file:///var/www/vendor/symfony/http-kernel/HttpKernel.php#L76) {▶}
    [/var/www/vendor/symfony/http-kernel/HttpCache/SubRequestHandler.php:86](file:///var/www/vendor/symfony/http-kernel/HttpCache/SubRequestHandler.php#L86) {▶}
    [/var/www/vendor/symfony/http-kernel/Fragment/InlineFragmentRenderer.php:78](file:///var/www/vendor/symfony/http-kernel/Fragment/InlineFragmentRenderer.php#L78) {▶}
    [/var/www/vendor/symfony/http-kernel/Fragment/FragmentHandler.php:83](file:///var/www/vendor/symfony/http-kernel/Fragment/FragmentHandler.php#L83) {▶}
    [/var/www/vendor/symfony/http-kernel/DependencyInjection/LazyLoadingFragmentHandler.php:47](file:///var/www/vendor/symfony/http-kernel/DependencyInjection/LazyLoadingFragmentHandler.php#L47) {▶}
    [/var/www/vendor/symfony/twig-bridge/Extension/HttpKernelRuntime.php:44](file:///var/www/vendor/symfony/twig-bridge/Extension/HttpKernelRuntime.php#L44) {▶}
    [/var/www/vendor/twig/twig/src/Environment.php:360](file:///var/www/vendor/twig/twig/src/Environment.php#L360) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:171](file:///var/www/vendor/twig/twig/src/Template.php#L171) {▶}
    [/var/www/vendor/twig/twig/src/Environment.php:360](file:///var/www/vendor/twig/twig/src/Environment.php#L360) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:394](file:///var/www/vendor/twig/twig/src/Template.php#L394) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:367](file:///var/www/vendor/twig/twig/src/Template.php#L367) {▶}
    [/var/www/vendor/twig/twig/src/Environment.php:360](file:///var/www/vendor/twig/twig/src/Environment.php#L360) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:394](file:///var/www/vendor/twig/twig/src/Template.php#L394) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:367](file:///var/www/vendor/twig/twig/src/Template.php#L367) {▶}
    [/var/www/vendor/twig/twig/src/Template.php:379](file:///var/www/vendor/twig/twig/src/Template.php#L379) {▶}
    [/var/www/vendor/twig/twig/src/TemplateWrapper.php:38](file:///var/www/vendor/twig/twig/src/TemplateWrapper.php#L38) {▶}
    [/var/www/vendor/twig/twig/src/Environment.php:280](file:///var/www/vendor/twig/twig/src/Environment.php#L280) {▶}
    [/var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php:448](file:///var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php#L448) {▶}
    [/var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php:453](file:///var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php#L453) {▶}
    [/var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php:253](file:///var/www/vendor/symfony/framework-bundle/Controller/AbstractController.php#L253) {▶}
    [/var/www/src/Controller/LandingController.php:84](file:///var/www/src/Controller/LandingController.php#L84) {▶}
    [/var/www/vendor/symfony/http-kernel/HttpKernel.php:181](file:///var/www/vendor/symfony/http-kernel/HttpKernel.php#L181) {▶}
    [/var/www/vendor/symfony/http-kernel/HttpKernel.php:76](file:///var/www/vendor/symfony/http-kernel/HttpKernel.php#L76) {▶}
    [/var/www/vendor/symfony/http-kernel/Kernel.php:197](file:///var/www/vendor/symfony/http-kernel/Kernel.php#L197) {▶}
    [/var/www/vendor/symfony/runtime/Runner/Symfony/HttpKernelRunner.php:35](file:///var/www/vendor/symfony/runtime/Runner/Symfony/HttpKernelRunner.php#L35) {▶}
    [/var/www/vendor/autoload_runtime.php:29](file:///var/www/vendor/autoload_runtime.php#L29) {▶}
    [/var/www/public/index.php:5](file:///var/www/public/index.php#L5) {▶}
```

**Comment by nicolas-grekas** [comment]
#52797 should fix this, isn't it?

**Comment by dmaicher** [comment]
@nicolas-grekas yes that solves it :rocket: thanks a lot

**Comment by mbabker** [comment]
Just another confirmation the issue is fixed after the upgrade from 6.4 RC to stable.

**Comment by raziel057** [comment]
While updating to SF6.4 I got the same error and can confirm that https://github.com/symfony/symfony/pull/52797 fix the issue.

![image](https://github.com/symfony/symfony/assets/652505/c3344cc4-d72d-4732-9b29-267cb85cc421)

**Comment by janklan** [comment]
> Just another confirmation the issue is fixed after the upgrade from 6.4 RC to stable.

I tripped this error after updating to the latest 6.4. RC was never on my machine. Is it fixed in 6.4.0, or not? https://github.com/symfony/symfony/pull/52797 doesn't seem to be part of 6.4.x.

I'm confused. 

Either way, adding a conflict of `"symfony/var-exporter": ">=6.3.9,<=6.4.0"` to my `composer.json` fixed the problem without forcing me to revert all packages back to 6.3.x.

Thank you Symfony maintainers for making this possible through zealous adherence to your BC promise.

**Comment by mbabker** [comment]
> > Just another confirmation the issue is fixed after the upgrade from 6.4 RC to stable.
> 
> I tripped this error after updating to the latest 6.4. RC was never on my machine. Is it fixed in 6.4.0, or not? #52797 doesn't seem to be part of 6.4.x.

I meant it worked fine on RC but I hit it on stable and the PR fixed it.  No idea why I couldn't communicate that well this morning 😅

**Comment by talha-core** [comment]
After #52797 fix, Now I am getting: `Cannot access uninitialized non-nullable property` on `vendor/symfony/var-exporter/LazyGhostTrait.php#L211`, Is this Issue is still worked on ?

**Comment by nicolas-grekas** [comment]
@talha-core can you please share a reproducer? And double check that the issue is not on your side (accessing an uninitialized property? - the stack trace should help)

**Comment by talha-core** [comment]
@nicolas-grekas  I don't know because, I was not getting this error on Symfony version 6.3.8, But when I upgraded to 6.4.0, I got these errors.

Here is a reproduced:
**Entity:**
```php
#[ORM\Entity(ArticleCombineRepository::class)]
class ArticleCombine
{
    /** @var Collection<int, ArticlePrice> */
    #[ORM\OneToMany(targetEntity: ArticlePrice::class, mappedBy: 'articleCombine', cascade: ['persist'])]
    #[ORM\OrderBy(['type' => 'ASC', 'qty' => 'ASC', 'id' => 'ASC'])]
    private Collection $articlePriceOptions;

    public function __construct()
    {
        $this->articlePriceOptions = new ArrayCollection();
    }

    public function getArticlePriceConsumerOptions(): mixed
    {
        return $this->articlePriceOptions->filter(
            function (ArticlePrice $articlePrice) {
                return $articlePrice->getType() == 'c';
            }
        );
    }
}
```

When I call `getArticlePriceConsumerOptions` It throws me that error.

**Comment by xabbuh** [comment]
@talha-core Are you sure that your constructor is being called? Otherwise the `$articlePriceOptions` wouldn't be initialised.

**Comment by talha-core** [comment]
> @talha-core Are you sure that your constructor is being called? Otherwise the `$articlePriceOptions` wouldn't be initialised.

Why it matters in Symfony 6.4 ? The entity is already persisted in Database. I am just fetching `ArticleCombine` from db and using this function.. I am confused.

**Comment by nicolas-grekas** [comment]
@talha-core Do you think you could prepare a small app reproducing the issue? That'd help a lot. Thanks!

**Comment by webda2l** [comment]
@talha-core Got the same error this morning, a clear cache fixed the issue.

Not exactly the same error after checking in fact. In my case it was:

> Typed property Symfony\Component\ErrorHandler\ErrorRenderer\FileLinkFormatter::$fileLinkFormat must not be accessed before initialization

when consult a profiling page.

**Comment by dkarlovi** [comment]
I can confirm this on 6.4.0 with clean cache, confirmed fixed by https://github.com/symfony/symfony/pull/52797, it works for me now.

**Comment by nicolas-grekas** [comment]
@talha-core does clearing the cache fix the issue you're having?

**Comment by shakaran** [comment]
When it will be backported to 7.0.0?

**Comment by talha-core** [comment]
@nicolas-grekas , I did, since running `composer install` clears the cache as well... But I didn't do it manually. 

I will try to recreate it when I will get the time. Thanks. :(

**Comment by Lustmored** [comment]
UPDATE: The following is fixed by #52797 

I just ran into this issue upon upgrading to 6.4.0 (`symfony/var-exporter` on both `7.0.0` and `6.4.0` act the same). I have removed `var/cache` and cleared the cache, but I still got:
```
An exception has been thrown during the rendering of a template ("Warning: Undefined array key 3") in "layout/article/_room.html.twig".
```

Which can be tracked down to those lines:
```
// Work around php/php-src#12695
                    $property = $propertyScopes[null === $scope ? $name : "\0$scope\0$name"][3]
                        ?? (Hydrator::$propertyScopes[$this::class] = Hydrator::getPropertyScopes($this::class))[3];
```

This happens to me on *some* (but not all) Doctrine entities. I *think* it happens on an entity property typed as integer with a value of "0". The line that triggers the problem looks like this (it has a nullsafe in it, so maybe that's the problem?):

```
return $this->difficulties[$room->getDifficulty()] ?? $this->difficulties[0];
```

Where `$room->getDifficulty()` throws.

**Comment by Lustmored** [comment]
But I can confirm my problem gets fixed with #52797 - sorry for the noise, I should've checked that in the first place
